proposals/nnnn-span-access-shared-contiguous-storage.md

```diff
@@ -284,114 +110,65 @@ extension Span where Element: ~Copyable & ~Escapable {
-  ///   `maxLength` must be greater than or equal to zero.
-  /// - Returns: A span with at most `maxLength` elements.
-  public func extracting(first maxLength: Int) -> Self
+  /// - Parameters:
+  ///   - index: an index to validate
+  /// - Returns: true if `index` is a valid index
+  public func boundsContain(_ index: Int) -> Bool
 
-  /// Returns a span over all but the given number of trailing elements.
-  ///
-  /// If the number of elements to drop exceeds the number of elements in
-  /// the span, the result is an empty span.
-  ///
-  /// The returned span's first item is always at offset 0; unlike buffer
-  /// slices, extracted spans do not share their indices with the
-  /// span from which they are extracted.
+  /// Traps if `index` is not a valid offset into this `Span`
   ///
-  /// - Parameter k: The number of elements to drop off the end of
-  ///   the span. `k` must be greater than or equal to zero.
-  /// - Returns: A span leaving off the specified number of elements at the end.
-  public func extracting(droppingLast k: Int) -> Self
+  /// - Parameters:
+  ///   - index: an index to validate
+  public func boundsPrecondition(_ index: Int)
 
-  /// Returns a span containing the final elements of the span,
-  /// up to the given maximum length.
-  ///
-  /// If the maximum length exceeds the length of this span,
-  /// the result contains all the elements.
+  /// Return true if `indices` is a valid range of offsets into this `Span`
   ///
-  /// The returned span's first item is always at offset 0; unlike buffer
-  /// slices, extracted spans do not share their indices with the
-  /// span from which they are extracted.
-  ///
-  /// - Parameter maxLength: The maximum number of elements to return.
-  ///   `maxLength` must be greater than or equal to zero.
-  /// - Returns: A span with at most `maxLength` elements.
-  public func extracting(last maxLength: Int) -> Self
+  /// - Parameters:
+  ///   - indices: a range of indices to validate
+  /// - Returns: true if `indices` is a valid range of indices
+  public func boundsContain(_ indices: Range<Int>) -> Bool
 
-  /// Returns a span over all but the given number of initial elements.
-  ///
-  /// If the number of elements to drop exceeds the number of elements in
-  /// the span, the result is an empty span.
+  /// Traps if `indices` is not a valid range of offsets into this `Span`
   ///
-  /// The returned span's first item is always at offset 0; unlike buffer
-  /// slices, extracted spans do not share their indices with the
-  /// span from which they are extracted.
-  ///
-  /// - Parameter k: The number of elements to drop from the beginning of
-  ///   the span. `k` must be greater than or equal to zero.
-  /// - Returns: A span starting after the specified number of elements.
-  public func extracting(droppingFirst k: Int) -> Self
+  /// - Parameters:
+  ///   - indices: a range of indices to validate
+  public func boundsPrecondition(_ indices: Range<Int>)
```

Comment:
`span.boundsPrecondition(range)` is just `precondition(span.boundsContain(range))`, right? Do we really need/want sugar for this?

Re: naming, I think things get somewhat easier if we drop the precondition func, then we just have `boundsContain`, whose naming isn't great because it's kind of oblique. The question we're really asking is "are the indices in this range valid for this span?". Something like `span.indicesContain(range)`, though that then raises the question of "why isn't this just `span.indices.contains(range)`?" (probably performance?)

Reply:
If we have these, do we also need `ClosedRange` versions?

Reply:
Yes, boundsPrecondition's implementation is precondition(boundsContain(range)). There's maybe an argument to be made that the sugar hurts here, since it doesn't capture the correct line of code in the defaulted arguments of precondition().

I haven't needed the `ClosedRange` versions because the `RangeExpression`-taking functions end up forwarding to the usual `Range`-taking functions. Maybe we would want them. We could take out the sugared preconditions and add ClosedRange bounds-checking functions.